src/object/cli_obj.c

```diff
@@ -4910,6 +5193,91 @@ obj_retry_next_shard(struct dc_object *obj, struct obj_auxi_args *obj_auxi,
 	return 0;
 }
 
+/* some pre-process for cond_fetch -
+ * 1. create local TX if it is no in a user transaction - to ensure all the fetches on same epoch
+ *    chosen by first operation (reuse dtx_epoch_chosen related logic).
+ * 2. for multiple-akeys case, split obj task to multiple sub-tasks each for one akey. For this
+ *    case return 1 to indicate wait sub-tasks' completion.
+ */
+static int
+obj_cond_fetch_prep(tse_task_t *task, struct obj_auxi_args *obj_auxi)
+{
+	daos_obj_fetch_t	*args = dc_task_get_args(task);
+	d_list_t		*task_list = &obj_auxi->shard_task_head;
+	tse_task_t		*sub_task;
+	d_sg_list_t		*sgl;
+	bool			 per_akey = args->flags & DAOS_COND_PER_AKEY;
+	uint64_t		 fetch_flags;
+	uint32_t		 i;
+	int			 rc = 0;
+
+	if (!daos_handle_is_valid(obj_auxi->th)) {
+		D_ASSERT(!obj_auxi->tx_local);
+		D_ASSERT(!daos_handle_is_valid(args->th));
+		rc = dc_tx_local_open(obj_auxi->obj->cob_coh, 0, 0, &obj_auxi->th);
+		if (rc) {
+			D_ERROR("task %p "DF_OID" dc_tx_local_open failed "
+				DF_RC"\n", task, DP_OID(obj_auxi->obj->cob_md.omd_id),
+				DP_RC(rc));
+			return rc;
+		}
+		args->th = obj_auxi->th;
+		obj_auxi->tx_local = 1;
+	}
+
+	if (args->nr <= 1 || (args->flags & (DAOS_COND_AKEY_FETCH | DAOS_COND_PER_AKEY)) == 0)
+		return rc;
+
+	/* If cond_fetch include multiple akeys, splits the obj task to multiple sub-tasks, one for
+	 * each akey. Because -
+	 * 1. for each akey's cond_fetch if any shard returns 0 (exist) then the akey is exist.
+	 * 2. for multi-akeys' cond_fetch, should return non-exist if any akey non-exist.
+	 * Now one fetch request only with one return code. So creates one sub-task for each akey.
+	 */
```

Comment:
that need to change RPC format which is I want to avoid. or can change like that later to change RPC together with other reason.

Reply:
>that need to change RPC format which cause interoperability issue which is I want to avoid.

I thought the old server will not support checking exist anyway?  what is the interoperability issue you mean here?


> that need quite some extra detailed code, also unnecessary for normal fetch RPC.

TBH, I think adding new RPC format/OPC is easier than adding these flags and sub-task, given how complicate of client stack.  IMHO, sub_task is being overused somehow.

Reply:
"the old server will not support checking exist anyway"
old server already support that, current patch only add some EC client process.

"adding new RPC format/OPC is easier than adding these flags and sub-task"
new RPC format cause interoperability issue I want to avoid, let's finish the functionality based on current RPC first. It seems no enough reason to change RPC if only for this purpose.